### PR TITLE
Plugins: Introduce new Flame graph panel

### PR DESCRIPTION
**Pull request by joey-grafana**
**What this PR does / why we need it**:

This PR adds a flame graph to Grafana.

**Special notes for your reviewer**:

- Feature toggle: `flameGraph`
- Will also need to add the new visualization type [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/frame_meta.go#L49). Update: PR [here](https://github.com/grafana/grafana-plugin-sdk-go/pull/546).

### CODE REVIEW

File: public/app/features/explore/Explore.tsx [hunk 1]
```diff
@@ -297,6 +310,15 @@ export class Explore extends React.PureComponent<Props, ExploreState> {
 
   memoizedGetNodeGraphDataFrames = memoizeOne(getNodeGraphDataFrames);
 
+  renderFlameGraphPanel() {
+    const { queryResponse } = this.props;
+    const dataFrames = queryResponse.series.filter(
```
Comment (aocenas) [inline public/app/features/explore/Explore.tsx:315]: Shouldn't we use `flameGraphFrames` here? Seems bit weird to populate them but then not use them and filter again here.

File: public/app/features/explore/utils/decorators.ts [hunk 2]
```diff
@@ -48,6 +49,9 @@ export const decorateWithFrameTypeMetadata = (data: PanelData): ExplorePanelData
       case 'nodeGraph':
         nodeGraphFrames.push(frame);
         break;
+      case 'flamegraph':
```
Comment (aocenas) [inline public/app/features/explore/utils/decorators.ts:52]: What about guarding with the flag here? That way users will at least get a table, which I think is a bit better (for knowing what is going on) then getting No Data when you query something that wants flamegraph.

**Comment by aocenas** [comment]
Looks good. Still needs a guard for the dashboard visualization selection but apart from that and those 2 comments this looks solid.

**Comment by joey-grafana** [comment]
@aocenas all comments addressed - thank you for the feedback.

Thanks again @wbrowne for suggesting a tidy (and temporary) solution for hiding the flame graph if its feature toggle is not enabled.

Screenshot 2022-10-06 at 13 30 58

**Review by aocenas — APPROVED** [review]
Nice work @joey-grafana lgmt

File: public/app/plugins/panel/flamegraph/plugin.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+  "type": "panel",
+  "name": "Flame Graph",
+  "id": "flamegraph",
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply public/app/plugins/panel/flamegraph/plugin.json:4]: Actually.. You can probably forget that as it's better to be consistent with what we have right now

**Review by wbrowne — APPROVED** [review]
LGTM 🚀

File: public/app/plugins/panel/flamegraph/plugin.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdboyer) [reply public/app/plugins/panel/flamegraph/plugin.json:4]: agree, consistency with what we have right now means that whatever solution we come up with won't need exception cases for the already half-migrated things :)

File: public/app/plugins/panel/flamegraph/components/FlameGraph/rendering.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  ctx.beginPath();
+  ctx.rect(rect.x + (rect.collapsed ? 0 : BAR_BORDER_WIDTH), rect.y, rect.width, rect.height);
+
+  //  / (rangeMax - rangeMin) here so when you click a bar it will adjust the top (clicked)bar to the most 'intense' color
+  const intensity = Math.min(1, rect.ticks / totalTicks / (rangeMax - rangeMin));
+  const h = 50 - 50 * intensity;
+  const l = 65 + 7 * intensity;
+
+  const name = rect.label;
+  const queryResult = query && fuzzyMatch(name.toLowerCase(), query.toLowerCase()).found;
```
Comment (gelicia) [inline public/app/plugins/panel/flamegraph/components/FlameGraph/rendering.ts:96]: I feel like with a search this lenient we may want to bold the highlighted ranges fuzzyMatch returns so people understand why things are being highlighted.

**Review by gelicia — APPROVED** [review]
I did my best code review for a 42k line PR ;) . I can see little things to be tweaked here and there about it but that can be figured out later (let me know if there's a followup ticket to track little UI suggestions). I think utilizing the ranges fuzzyMatch returns will be a bit higher of a priority but this all looks like a good pass at this.

**Review by cyriltovena — APPROVED** [review]
LGTM

File: public/app/plugins/panel/flamegraph/components/FlameGraph/rendering.ts (hunk 4 above)
Comment (joey-grafana) [reply public/app/plugins/panel/flamegraph/components/FlameGraph/rendering.ts:96]: 100% - this is something we discussed during the week. I will be updating that next week :)

**Pull request merged**